## Assignment 1

In [13]:
import tensorflow as tf
import matplotlib.pyplot as plt
import numpy as np
import random as rn

## Part One:

* How many total trainable parameters does this model have?
  * How many weights?->> 101632
  * How many biases?->> 138
* How accurate was this model on the validation data after 10 epochs?
  * How different was the models performance on the training data?->>> 0.98
  * How different was the models performance on the test data?->>> 0.97
* About how long did each epoch take?->>> 2seconds and 2milliseconds

In [8]:
print("tensorflow version"+tf.__version__)

tensorflow version2.3.1


In [10]:
mnist=tf.keras.datasets.mnist
(train_images,train_labels),(test_images,test_labels)=mnist.load_data()

In [14]:
train_images=(np.expand_dims(train_images,axis=-1)/255).astype(np.float32)
test_images = (np.expand_dims(test_images, axis=-1)/255.).astype(np.float32)

In [16]:
train_labels=(train_labels).astype(np.int64)
test_labels=(test_labels).astype(np.int64)

In [27]:
def model_mnist():
    model= tf.keras.Sequential([
      tf.keras.layers.Flatten(),
      tf.keras.layers.Dense(128, activation= tf.nn.relu),
      tf.keras.layers.Dense(10,activation=tf.nn.softmax)])
    return(model)

In [28]:
m1=model_mnist()

In [29]:
m1.compile(optimizer=tf.keras.optimizers.SGD(learning_rate=1e-1), 
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [31]:
BATCH_SIZE = 64
EPOCHS = 10
m1.fit(train_images, train_labels, batch_size=BATCH_SIZE, epochs=EPOCHS)

Epoch 1/10
938/938 [==============================] - 3s 4ms/step - loss: 0.3753 - accuracy: 0.8958
Epoch 2/10
938/938 [==============================] - 2s 2ms/step - loss: 0.2031 - accuracy: 0.9420
Epoch 3/10
938/938 [==============================] - 2s 2ms/step - loss: 0.1545 - accuracy: 0.9561
Epoch 4/10
938/938 [==============================] - 2s 2ms/step - loss: 0.1255 - accuracy: 0.9639
Epoch 5/10
938/938 [==============================] - 2s 3ms/step - loss: 0.1053 - accuracy: 0.9703
Epoch 6/10
938/938 [==============================] - 2s 3ms/step - loss: 0.0909 - accuracy: 0.9743
Epoch 7/10
938/938 [==============================] - 2s 2ms/step - loss: 0.0801 - accuracy: 0.9775
Epoch 8/10
938/938 [==============================] - 2s 2ms/step - loss: 0.0716 - accuracy: 0.9797
Epoch 9/10
938/938 [==============================] - 2s 2ms/step - loss: 0.0643 - accuracy: 0.9815
Epoch 10/10
938/938 [==============================] - 2s 2ms/step - loss: 0.0586 - accuracy: 0.9833

In [33]:
test_loss, test_acc = m1.evaluate(test_images,test_labels)
print('Test accuracy:', test_acc)

313/313 [==============================] - 1s 2ms/step - loss: 0.0791 - accuracy: 0.9748
Test accuracy: 0.9747999906539917


## Part Two:

Use Keras to build 3 networks, each with at least 10 hidden layers such that:

* The first model has fewer than 10 nodes per layer.
* The second model has between 10-50 nodes per layer.
* The third model has between 50-100 nodes per layer.


In [34]:
def model_mnist(n_nodes):
    model= tf.keras.Sequential([
      tf.keras.layers.Flatten(),
      tf.keras.layers.Dense(n_nodes, activation= tf.nn.relu),
      tf.keras.layers.Dense(n_nodes, activation= tf.nn.relu),
      tf.keras.layers.Dense(n_nodes, activation= tf.nn.relu),
      tf.keras.layers.Dense(n_nodes, activation= tf.nn.relu),
      tf.keras.layers.Dense(n_nodes, activation= tf.nn.relu),
      tf.keras.layers.Dense(n_nodes, activation= tf.nn.relu),
      tf.keras.layers.Dense(n_nodes, activation= tf.nn.relu),
      tf.keras.layers.Dense(n_nodes, activation= tf.nn.relu),
      tf.keras.layers.Dense(n_nodes, activation= tf.nn.relu),
      tf.keras.layers.Dense(n_nodes,activation=tf.nn.softmax)])
    return(model)

In [35]:
m1=model_mnist(10)
m2=model_mnist(50)
m3=model_mnist(100)
m1.compile(optimizer=tf.keras.optimizers.SGD(learning_rate=1e-1), 
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
BATCH_SIZE = 64
EPOCHS = 10
m1.fit(train_images, train_labels, batch_size=BATCH_SIZE, epochs=EPOCHS)
test_loss, test_acc = m1.evaluate(test_images,test_labels)
print('Test accuracy:', test_acc)

Epoch 1/10
938/938 [==============================] - 3s 3ms/step - loss: 1.4957 - accuracy: 0.4323
Epoch 2/10
938/938 [==============================] - 2s 2ms/step - loss: 0.7834 - accuracy: 0.7152
Epoch 3/10
938/938 [==============================] - 2s 3ms/step - loss: 0.6437 - accuracy: 0.7869
Epoch 4/10
938/938 [==============================] - 2s 2ms/step - loss: 0.4761 - accuracy: 0.8784
Epoch 5/10
938/938 [==============================] - 2s 3ms/step - loss: 0.3886 - accuracy: 0.8996
Epoch 6/10
938/938 [==============================] - 2s 2ms/step - loss: 0.3637 - accuracy: 0.9053
Epoch 7/10
938/938 [==============================] - 2s 2ms/step - loss: 0.3305 - accuracy: 0.9138
Epoch 8/10
938/938 [==============================] - 3s 3ms/step - loss: 0.3105 - accuracy: 0.9191
Epoch 9/10
938/938 [==============================] - 3s 3ms/step - loss: 0.2949 - accuracy: 0.9230
Epoch 10/10
313/313 [==============================] - 1s 2ms/step - loss: 0.2824 - accuracy: 0.9252

In [36]:
m2.compile(optimizer=tf.keras.optimizers.SGD(learning_rate=1e-1), 
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
BATCH_SIZE = 64
EPOCHS = 10
m2.fit(train_images, train_labels, batch_size=BATCH_SIZE, epochs=EPOCHS)
test_loss, test_acc = m2.evaluate(test_images,test_labels)
print('Test accuracy:', test_acc)

Epoch 1/10
938/938 [==============================] - 3s 3ms/step - loss: 1.0324 - accuracy: 0.6603
Epoch 2/10
938/938 [==============================] - 3s 3ms/step - loss: 0.2510 - accuracy: 0.9321
Epoch 3/10
938/938 [==============================] - 3s 4ms/step - loss: 0.1737 - accuracy: 0.9517
Epoch 4/10
938/938 [==============================] - 3s 4ms/step - loss: 0.1378 - accuracy: 0.9610
Epoch 5/10
938/938 [==============================] - 3s 3ms/step - loss: 0.1166 - accuracy: 0.9669
Epoch 6/10
938/938 [==============================] - 3s 4ms/step - loss: 0.1024 - accuracy: 0.9699
Epoch 7/10
938/938 [==============================] - 3s 3ms/step - loss: 0.0912 - accuracy: 0.9739
Epoch 8/10
938/938 [==============================] - 3s 3ms/step - loss: 0.0810 - accuracy: 0.9769
Epoch 9/10
938/938 [==============================] - 3s 3ms/step - loss: 0.0742 - accuracy: 0.9782
Epoch 10/10
313/313 [==============================] - 1s 3ms/step - loss: 0.1300 - accuracy: 0.9653

In [37]:
m3.compile(optimizer=tf.keras.optimizers.SGD(learning_rate=1e-1), 
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
BATCH_SIZE = 64
EPOCHS = 10
m3.fit(train_images, train_labels, batch_size=BATCH_SIZE, epochs=EPOCHS)
test_loss, test_acc = m3.evaluate(test_images,test_labels)
print('Test accuracy:', test_acc)

Epoch 1/10
938/938 [==============================] - 4s 5ms/step - loss: 0.9713 - accuracy: 0.6771
Epoch 2/10
938/938 [==============================] - 4s 4ms/step - loss: 0.1852 - accuracy: 0.9471
Epoch 3/10
938/938 [==============================] - 4s 4ms/step - loss: 0.1223 - accuracy: 0.9649
Epoch 4/10
938/938 [==============================] - 4s 5ms/step - loss: 0.0964 - accuracy: 0.9724
Epoch 5/10
938/938 [==============================] - 4s 5ms/step - loss: 0.0782 - accuracy: 0.9773
Epoch 6/10
938/938 [==============================] - 4s 5ms/step - loss: 0.0677 - accuracy: 0.9805
Epoch 7/10
938/938 [==============================] - 4s 5ms/step - loss: 0.0563 - accuracy: 0.9837
Epoch 8/10
938/938 [==============================] - 4s 5ms/step - loss: 0.0489 - accuracy: 0.9856
Epoch 9/10
938/938 [==============================] - 4s 5ms/step - loss: 0.0421 - accuracy: 0.9870
Epoch 10/10
313/313 [==============================] - 1s 4ms/step - loss: 0.0840 - accuracy: 0.9768

* Did any of these models achieve better than 20% accuracy on validation or test data?
  ->> yes, all of them
* How many total trainable parameters do each of these models have?
    ->> ~n_nodes^10 + n_nodes*10
* Is there a clear correlation between number of trainable parameters and accuracy?
    ->> more nodes ie. parameters there is increase in test accuracy

## Part Three:

In [38]:
m3.compile(optimizer=tf.keras.optimizers.SGD(learning_rate=1e-1), 
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
BATCH_SIZE = 64
EPOCHS = 10
m3.fit(train_images, train_labels, batch_size=BATCH_SIZE, epochs=EPOCHS)
test_loss, test_acc = m3.evaluate(test_images,test_labels)
print('Test accuracy:', test_acc)

Epoch 1/10
938/938 [==============================] - 5s 5ms/step - loss: 0.0356 - accuracy: 0.9895
Epoch 2/10
938/938 [==============================] - 4s 5ms/step - loss: 0.0300 - accuracy: 0.9906
Epoch 3/10
938/938 [==============================] - 4s 5ms/step - loss: 0.0267 - accuracy: 0.9918
Epoch 4/10
938/938 [==============================] - 4s 5ms/step - loss: 0.0268 - accuracy: 0.9913
Epoch 5/10
938/938 [==============================] - 5s 5ms/step - loss: 0.0225 - accuracy: 0.9931
Epoch 6/10
938/938 [==============================] - 5s 5ms/step - loss: 0.0166 - accuracy: 0.9947
Epoch 7/10
938/938 [==============================] - 4s 4ms/step - loss: 0.0183 - accuracy: 0.9947
Epoch 8/10
938/938 [==============================] - 4s 5ms/step - loss: 0.0172 - accuracy: 0.9950
Epoch 9/10
938/938 [==============================] - 4s 5ms/step - loss: 0.0154 - accuracy: 0.9955
Epoch 10/10
313/313 [==============================] - 1s 2ms/step - loss: 0.1143 - accuracy: 0.9745

Build a network with at least 3 hidden layers that achieves better than 92% accuracy on validation and test data.
You may need to train for more than 10 epochs to achieve this result.

* Compare your best results to the result you got in part one:
  * Which network achieved the best accuracy on test data after training?
  ->> the NN in part 1 had performed better as it had test accuracy 0.974799 with last one at 0.9745
  * Did the networks train for a similar number of epochs?
  ->> same no. of epochs 10